File: bot/db/sqlite.py
import sqlite3
import logging
import json
from datetime import datetime, timedelta
from typing import Any, List, Dict, Optional
from config import get_config

logger = logging.getLogger(__name__)


# Register adapters for datetime
def adapt_datetime(dt: datetime) -> str:
    return dt.isoformat()


def convert_datetime(s: bytes) -> Optional[datetime]:
    try:
        return datetime.fromisoformat(s.decode())
    except ValueError, TypeError:
        return None


sqlite3.register_adapter(datetime, adapt_datetime)
sqlite3.register_converter("DATETIME", convert_datetime)


# pylint: disable=too-many-public-methods
class BotDB:
    def __init__(self, db_path: Optional[str] = None):
        if db_path is None:
            conf = get_config()
            self.db_path = conf["SQLITE_DB_PATH"]
        else:
            self.db_path = db_path
        self._init_db()

    def _get_conn(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self.db_path, detect_types=sqlite3.PARSE_DECLTYPES)
        conn.row_factory = sqlite3.Row
        return conn

    def _init_db(self) -> None:
        logger.info("Initializing SQLite database at %s", self.db_path)
        with self._get_conn() as conn:
            # Trusted Users
            conn.execute("""
                CREATE TABLE IF NOT EXISTS trusted_users (
                    user_id INTEGER PRIMARY KEY,
                    "by" INTEGER,
                    datetime DATETIME
                )
            """)
            # Buktopuha Players
            conn.execute("""
                CREATE TABLE IF NOT EXISTS buktopuha_players (
                    user_id INTEGER PRIMARY KEY,
                    meta TEXT,
                    game_counter INTEGER DEFAULT 0,
                    win_counter INTEGER DEFAULT 0,
                    total_score INTEGER DEFAULT 0,
                    created_at DATETIME,
                    updated_at DATETIME
                )
            """)
            # Towel Quarantine
            conn.execute("""
                CREATE TABLE IF NOT EXISTS towel_quarantine (
                    user_id INTEGER PRIMARY KEY,
                    rel_messages TEXT,
                    datetime DATETIME
                )
            """)
            # Since Topics
            conn.execute("""
                CREATE TABLE IF NOT EXISTS since_topics (
                    topic TEXT PRIMARY KEY,
                    since_datetime DATETIME,
                    count INTEGER DEFAULT 0
                )
            """)
            # Roll Hussars
            conn.execute("""
                CREATE TABLE IF NOT EXISTS roll_hussars (
                    user_id INTEGER PRIMARY KEY,
                    meta TEXT,
                    shot_counter INTEGER DEFAULT 0,
                    miss_counter INTEGER DEFAULT 0,
                    dead_counter INTEGER DEFAULT 0,
                    total_time_in_club INTEGER DEFAULT 0,
                    first_shot DATETIME,
                    last_shot DATETIME
                )
            """)
            # Prism Words
            conn.execute("""
                CREATE TABLE IF NOT EXISTS prism_words (
                    word TEXT PRIMARY KEY,
                    count INTEGER DEFAULT 0,
                    last_use DATETIME
                )
            """)
            # Peninsula Users
            conn.execute("""
                CREATE TABLE IF NOT EXISTS peninsula_users (
                    user_id INTEGER PRIMARY KEY,
                    meta TEXT
                )
            """)
            # AOC Data
            conn.execute("""
                CREATE TABLE IF NOT EXISTS aoc (
                    id INTEGER PRIMARY KEY CHECK (id = 0),
                    data TEXT
                )
            """)
            conn.commit()

    # --- Trusted Users ---
    def get_trusted_user(self, user_id: int) -> Optional[Dict[str, Any]]:
        row = self.fetchone("SELECT * FROM trusted_users WHERE user_id = ?", (user_id,))
        return dict(row) if row else None

    def trust_user(self, user_id: int, admin_id: int) -> None:
        self.execute(
            'INSERT OR REPLACE INTO trusted_users (user_id, "by", datetime) VALUES (?, ?, ?)',
            (user_id, admin_id, datetime.now()),
        )

    def untrust_user(self, user_id: int) -> None:
        self.execute("DELETE FROM trusted_users WHERE user_id = ?", (user_id,))

    def is_user_trusted(self, user_id: int | str) -> bool:
        row = self.fetchone("SELECT 1 FROM trusted_users WHERE user_id = ?", (user_id,))
        return row is not None

    # --- Buktopuha ---
    def get_all_buktopuha_players(self) -> List[Dict[str, Any]]:
        rows = self.fetchall(
            "SELECT * FROM buktopuha_players ORDER BY win_counter DESC"
        )
        res: List[Dict[str, Any]] = []
        for r in rows:
            d = dict(r)
            d["_id"] = d.pop("user_id")
            d["meta"] = json.loads(d["meta"])
            res.append(d)
        return res

    def find_buktopuha_player(self, user_id: int) -> Optional[Dict[str, Any]]:
        row = self.fetchone(
            "SELECT * FROM buktopuha_players WHERE user_id = ?", (user_id,)
        )
        if row:
            d = dict(row)
            d["_id"] = d.pop("user_id")
            d["meta"] = json.loads(d["meta"])
            return d
        return None

    def add_buktopuha_player(
        self, user_id: int, user_meta: Dict[str, Any], score: int = 0
    ) -> None:
        now = datetime.now()
        game_inc = 1 if score == 0 else 0
        win_inc = 1 if score > 0 else 0
        self.execute(
            "INSERT INTO buktopuha_players (user_id, meta, game_counter, win_counter, total_score, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
            (user_id, json.dumps(user_meta), game_inc, win_inc, score, now, now),
        )

    def inc_buktopuha_game_counter(self, user_id: int) -> None:
        self.execute(
            "UPDATE buktopuha_players SET game_counter = game_counter + 1, updated_at = ? WHERE user_id = ?",
            (datetime.now(), user_id),
        )

    def inc_buktopuha_win(self, user_id: int, score: int) -> None:
        self.execute(
            "UPDATE buktopuha_players SET win_counter = win_counter + 1, total_score = total_score + ?, updated_at = ? WHERE user_id = ?",
            (score, datetime.now(), user_id),
        )

    def remove_buktopuha_player(self, user_id: int) -> None:
        self.execute("DELETE FROM buktopuha_players WHERE user_id = ?", (user_id,))

    def remove_all_buktopuha_players(self) -> None:
        self.execute("DELETE FROM buktopuha_players")

    # --- Towel Quarantine ---
    def add_quarantine_user(self, user_id: int, quarantine_time_min: int) -> None:
        if self.find_quarantine_user(user_id) is not None:
            return
        self.execute(
            "INSERT INTO towel_quarantine (user_id, rel_messages, datetime) VALUES (?, ?, ?)",
            (
                user_id,
                json.dumps([]),
                datetime.now() + timedelta(minutes=quarantine_time_min),
            ),
        )

    def find_quarantine_user(self, user_id: int) -> Optional[Dict[str, Any]]:
        row = self.fetchone(
            "SELECT * FROM towel_quarantine WHERE user_id = ?", (user_id,)
        )
        if row:
            d = dict(row)
            d["_id"] = d.pop("user_id")
            d["rel_messages"] = json.loads(d["rel_messages"])
            return d
        return None

    def find_all_quarantine_users(self) -> List[Dict[str, Any]]:
        rows = self.fetchall("SELECT * FROM towel_quarantine")
        res: List[Dict[str, Any]] = []
        for r in rows:
            d = dict(r)
            d["_id"] = d.pop("user_id")
            d["rel_messages"] = json.loads(d["rel_messages"])
            res.append(d)
        return res

    def add_quarantine_rel_message(self, user_id: int, message_id: int) -> None:
        user = self.find_quarantine_user(user_id)
        if user:
            rel_messages = set(user["rel_messages"])
            rel_messages.add(message_id)
            self.execute(
                "UPDATE towel_quarantine SET rel_messages = ? WHERE user_id = ?",
                (json.dumps(list(rel_messages)), user_id),
            )

    def delete_quarantine_user(self, user_id: int) -> None:
        self.execute("DELETE FROM towel_quarantine WHERE user_id = ?", (user_id,))

    def delete_all_quarantine_users(self) -> None:
        self.execute("DELETE FROM towel_quarantine")

    # --- Since Topics ---
    def get_since_topic(self, topic: str) -> Dict[str, Any]:
        row = self.fetchone(
            "SELECT * FROM since_topics WHERE topic = ?", (topic.lower(),)
        )
        if row:
            return dict(row)
        return {"topic": topic.lower(), "since_datetime": datetime.now(), "count": 1}

    def update_since_topic(
        self, topic: str, since_datetime: datetime, count: int
    ) -> None:
        row = self.fetchone(
            "SELECT 1 FROM since_topics WHERE topic = ?", (topic.lower(),)
        )
        if row:
            self.execute(
                "UPDATE since_topics SET count = count + 1, since_datetime = ? WHERE topic = ?",
                (datetime.now(), topic.lower()),
            )
        else:
            self.execute(
                "INSERT INTO since_topics (topic, since_datetime, count) VALUES (?, ?, ?)",
                (topic.lower(), since_datetime, count),
            )

    def get_all_since_topics(self, limit: int) -> List[Dict[str, Any]]:
        rows = self.fetchall(
            "SELECT * FROM since_topics ORDER BY count DESC LIMIT ?", (limit,)
        )
        return [dict(r) for r in rows]

    # --- Roll Hussars ---
    def get_all_hussars(self) -> List[Dict[str, Any]]:
        rows = self.fetchall(
            "SELECT * FROM roll_hussars ORDER BY total_time_in_club DESC"
        )
        res: List[Dict[str, Any]] = []
        for r in rows:
            d = dict(r)
            d["_id"] = d.pop("user_id")
            d["meta"] = json.loads(d["meta"])
            res.append(d)
        return res

    def find_hussar(self, user_id: int) -> Optional[Dict[str, Any]]:
        row = self.fetchone("SELECT * FROM roll_hussars WHERE user_id = ?", (user_id,))
        if row:
            d = dict(row)
            d["_id"] = d.pop("user_id")
            d["meta"] = json.loads(d["meta"])
            return d
        return None

    def add_hussar(self, user_id: int, user_meta: Dict[str, Any]) -> None:
        now = datetime.now()
        self.execute(
            "INSERT INTO roll_hussars (user_id, meta, shot_counter, miss_counter, dead_counter, total_time_in_club, first_shot, last_shot) VALUES (?, ?, 0, 0, 0, 0, ?, ?)",
            (user_id, json.dumps(user_meta), now, now),
        )

    def hussar_dead(self, user_id: int, mute_min: int) -> None:
        self.execute(
            "UPDATE roll_hussars SET shot_counter = shot_counter + 1, dead_counter = dead_counter + 1, total_time_in_club = total_time_in_club + ?, last_shot = ? WHERE user_id = ?",
            (mute_min * 60, datetime.now(), user_id),
        )

    def hussar_miss(self, user_id: int) -> None:
        self.execute(
            "UPDATE roll_hussars SET shot_counter = shot_counter + 1, miss_counter = miss_counter + 1, last_shot = ? WHERE user_id = ?",
            (datetime.now(), user_id),
        )

    def remove_hussar(self, user_id: int) -> None:
        self.execute("DELETE FROM roll_hussars WHERE user_id = ?", (user_id,))

    def remove_all_hussars(self) -> None:
        self.execute("DELETE FROM roll_hussars")

    # --- Prism Words ---
    def add_prism_word(self, word: str) -> None:
        row = self.fetchone("SELECT 1 FROM prism_words WHERE word = ?", (word.lower(),))
        if row:
            self.execute(
                "UPDATE prism_words SET count = count + 1, last_use = ? WHERE word = ?",
                (datetime.now(), word.lower()),
            )
        else:
            self.execute(
                "INSERT INTO prism_words (word, count, last_use) VALUES (?, 1, ?)",
                (word.lower(), datetime.now()),
            )

    def get_all_prism_words(self) -> List[Dict[str, Any]]:
        rows = self.fetchall("SELECT * FROM prism_words ORDER BY count DESC")
        return [dict(r) for r in rows]

    # --- Peninsula Users ---
    def get_best_peninsulas(self, n: int = 10) -> List[Dict[str, Any]]:
        rows = self.fetchall(
            "SELECT * FROM peninsula_users ORDER BY user_id ASC LIMIT ?", (n,)
        )
        res: List[Dict[str, Any]] = []
        for r in rows:
            d = dict(r)
            d["_id"] = d.pop("user_id")
            d["meta"] = json.loads(d["meta"])
            res.append(d)
        return res

    def add_peninsula_user(self, user_id: int, user_meta: Dict[str, Any]) -> None:
        self.execute(
            "INSERT OR REPLACE INTO peninsula_users (user_id, meta) VALUES (?, ?)",
            (user_id, json.dumps(user_meta)),
        )

    # --- AOC Data ---
    def update_aoc_data(self, data: Dict[str, Any]) -> None:
        self.execute(
            "INSERT OR REPLACE INTO aoc (id, data) VALUES (0, ?)",
            (json.dumps(data),),
        )

    def get_aoc_data(self) -> Optional[Dict[str, Any]]:
        row = self.fetchone("SELECT data FROM aoc WHERE id = 0")
        if row:
            return json.loads(row["data"])
        return None

    def remove_all_aoc_data(self) -> None:
        self.execute("DELETE FROM aoc")

    # --- Generic helpers ---
    def execute(self, query: str, params: tuple[Any, ...] = ()) -> sqlite3.Cursor:
        with self._get_conn() as conn:
            return conn.execute(query, params)

    def fetchone(
        self, query: str, params: tuple[Any, ...] = ()
    ) -> Optional[sqlite3.Row]:
        with self._get_conn() as conn:
            return conn.execute(query, params).fetchone()

    def fetchall(self, query: str, params: tuple[Any, ...] = ()) -> List[sqlite3.Row]:
        with self._get_conn() as conn:
            return conn.execute(query, params).fetchall()


# Global instance
db = BotDB()

File: internal/db/sqlite.go
package db

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/jmoiron/sqlx"
	_ "modernc.org/sqlite" // SQLite driver
)

type DB struct {
	db *sqlx.DB
}

func New(dbPath string) (*DB, error) {
	conn, err := sqlx.Open("sqlite", dbPath+"?_pragma=journal_mode(WAL)&_pragma=busy_timeout(5000)")
	if err != nil {
		return nil, fmt.Errorf("open db: %w", err)
	}

	// SQLite: single writer, multiple readers
	conn.SetMaxOpenConns(1)

	d := &DB{db: conn}
	if err := d.init(); err != nil {
		return nil, fmt.Errorf("init db: %w", err)
	}
	return d, nil
}

func (d *DB) Close() error {
	return d.db.Close()
}

func (d *DB) init() error {
	schema := `
	CREATE TABLE IF NOT EXISTS trusted_users (
		user_id INTEGER PRIMARY KEY,
		added_by INTEGER,
		created_at TEXT
	);
	CREATE TABLE IF NOT EXISTS buktopuha_players (
		user_id INTEGER PRIMARY KEY,
		meta TEXT,
		game_counter INTEGER DEFAULT 0,
		win_counter INTEGER DEFAULT 0,
		total_score INTEGER DEFAULT 0,
		created_at TEXT,
		updated_at TEXT
	);
	CREATE TABLE IF NOT EXISTS towel_quarantine (
		user_id INTEGER PRIMARY KEY,
		rel_messages TEXT,
		expires_at TEXT
	);
	CREATE TABLE IF NOT EXISTS since_topics (
		topic TEXT PRIMARY KEY,
		since_datetime TEXT,
		count INTEGER DEFAULT 0
	);
	CREATE TABLE IF NOT EXISTS roll_hussars (
		user_id INTEGER PRIMARY KEY,
		meta TEXT,
		shot_counter INTEGER DEFAULT 0,
		miss_counter INTEGER DEFAULT 0,
		dead_counter INTEGER DEFAULT 0,
		total_time_in_club INTEGER DEFAULT 0,
		first_shot TEXT,
		last_shot TEXT
	);
	CREATE TABLE IF NOT EXISTS prism_words (
		word TEXT PRIMARY KEY,
		count INTEGER DEFAULT 0,
		last_use TEXT
	);
	CREATE TABLE IF NOT EXISTS peninsula_users (
		user_id INTEGER PRIMARY KEY,
		meta TEXT
	);
	CREATE TABLE IF NOT EXISTS aoc (
		id INTEGER PRIMARY KEY CHECK (id = 0),
		data TEXT
	);
	CREATE TABLE IF NOT EXISTS length_users (
		user_id INTEGER PRIMARY KEY,
		username TEXT,
		first_name TEXT,
		last_name TEXT,
		length INTEGER DEFAULT 0
	);`
	_, err := d.db.Exec(schema)
	return err
}

// --- Trusted Users ---

type TrustedUser struct {
	UserID    int64  `db:"user_id"`
	AddedBy   int64  `db:"added_by"`
	CreatedAt string `db:"created_at"`
}

func (d *DB) TrustUser(userID, adminID int64) error {
	_, err := d.db.Exec(
		`INSERT OR REPLACE INTO trusted_users (user_id, added_by, created_at) VALUES (?, ?, ?)`,
		userID, adminID, time.Now().Format(time.RFC3339),
	)
	return err
}

func (d *DB) UntrustUser(userID int64) error {
	_, err := d.db.Exec(`DELETE FROM trusted_users WHERE user_id = ?`, userID)
	return err
}

func (d *DB) IsUserTrusted(userID int64) (bool, error) {
	var count int
	err := d.db.Get(&count, `SELECT COUNT(1) FROM trusted_users WHERE user_id = ?`, userID)
	return count > 0, err
}

// --- Towel Quarantine ---

type QuarantineUser struct {
	UserID      int64  `db:"user_id"`
	RelMessages string `db:"rel_messages"` // JSON array of message IDs
	ExpiresAt   string `db:"expires_at"`
}

func (q *QuarantineUser) IsExpired() bool {
	t, err := time.Parse(time.RFC3339, q.ExpiresAt)
	if err != nil {
		return true
	}
	return time.Now().After(t)
}

func (q *QuarantineUser) MessageIDs() []int {
	var ids []int
	_ = json.Unmarshal([]byte(q.RelMessages), &ids)
	return ids
}

func (d *DB) AddQuarantineUser(userID int64, quarantineMinutes int) error {
	existing, _ := d.FindQuarantineUser(userID)
	if existing != nil {
		return nil
	}
	expiresAt := time.Now().Add(time.Duration(quarantineMinutes) * time.Minute).Format(time.RFC3339)
	_, err := d.db.Exec(
		`INSERT INTO towel_quarantine (user_id, rel_messages, expires_at) VALUES (?, ?, ?)`,
		userID, "[]", expiresAt,
	)
	return err
}

func (d *DB) FindQuarantineUser(userID int64) (*QuarantineUser, error) {
	var u QuarantineUser
	err := d.db.Get(&u, `SELECT * FROM towel_quarantine WHERE user_id = ?`, userID)
	if err != nil {
		return nil, err
	}
	return &u, nil
}

func (d *DB) FindAllQuarantineUsers() ([]QuarantineUser, error) {
	var users []QuarantineUser
	err := d.db.Select(&users, `SELECT * FROM towel_quarantine`)
	return users, err
}

func (d *DB) AddQuarantineRelMessage(userID int64, messageID int) error {
	u, err := d.FindQuarantineUser(userID)
	if err != nil {
		return err
	}
	ids := u.MessageIDs()
	// deduplicate
	for _, id := range ids {
		if id == messageID {
			return nil
		}
	}
	ids = append(ids, messageID)
	data, _ := json.Marshal(ids)
	_, err = d.db.Exec(`UPDATE towel_quarantine SET rel_messages = ? WHERE user_id = ?`, string(data), userID)
	return err
}

func (d *DB) DeleteQuarantineUser(userID int64) error {
	_, err := d.db.Exec(`DELETE FROM towel_quarantine WHERE user_id = ?`, userID)
	return err
}

func (d *DB) DeleteAllQuarantineUsers() error {
	_, err := d.db.Exec(`DELETE FROM towel_quarantine`)
	return err
}

// --- Roll Hussars ---

type Hussar struct {
	UserID          int64  `db:"user_id"`
	Meta            string `db:"meta"` // JSON
	ShotCounter     int    `db:"shot_counter"`
	MissCounter     int    `db:"miss_counter"`
	DeadCounter     int    `db:"dead_counter"`
	TotalTimeInClub int    `db:"total_time_in_club"` // seconds
	FirstShot       string `db:"first_shot"`
	LastShot        string `db:"last_shot"`
}

func (h *Hussar) Username() string {
	var meta map[string]any
	_ = json.Unmarshal([]byte(h.Meta), &meta)
	if username, ok := meta["username"].(string); ok && username != "" {
		return username
	}
	firstName, _ := meta["first_name"].(string)
	lastName, _ := meta["last_name"].(string)
	name := firstName
	if lastName != "" {
		if name != "" {
			name += " "
		}
		name += lastName
	}
	if name == "" {
		return "unknown"
	}
	return name
}

func (d *DB) GetAllHussars() ([]Hussar, error) {
	var hussars []Hussar
	err := d.db.Select(&hussars, `SELECT * FROM roll_hussars ORDER BY total_time_in_club DESC`)
	return hussars, err
}

func (d *DB) FindHussar(userID int64) (*Hussar, error) {
	var h Hussar
	err := d.db.Get(&h, `SELECT * FROM roll_hussars WHERE user_id = ?`, userID)
	if err != nil {
		return nil, err
	}
	return &h, nil
}

func (d *DB) AddHussar(userID int64, meta string) error {
	now := time.Now().Format(time.RFC3339)
	_, err := d.db.Exec(
		`INSERT INTO roll_hussars (user_id, meta, shot_counter, miss_counter, dead_counter, total_time_in_club, first_shot, last_shot) VALUES (?, ?, 0, 0, 0, 0, ?, ?)`,
		userID, meta, now, now,
	)
	return err
}

func (d *DB) HussarDead(userID int64, muteMinutes int) error {
	_, err := d.db.Exec(
		`UPDATE roll_hussars SET shot_counter = shot_counter + 1, dead_counter = dead_counter + 1, total_time_in_club = total_time_in_club + ?, last_shot = ? WHERE user_id = ?`,
		muteMinutes*60, time.Now().Format(time.RFC3339), userID,
	)
	return err
}

func (d *DB) HussarMiss(userID int64) error {
	_, err := d.db.Exec(
		`UPDATE roll_hussars SET shot_counter = shot_counter + 1, miss_counter = miss_counter + 1, last_shot = ? WHERE user_id = ?`,
		time.Now().Format(time.RFC3339), userID,
	)
	return err
}

func (d *DB) RemoveHussar(userID int64) error {
	_, err := d.db.Exec(`DELETE FROM roll_hussars WHERE user_id = ?`, userID)
	return err
}

func (d *DB) RemoveAllHussars() error {
	_, err := d.db.Exec(`DELETE FROM roll_hussars`)
	return err
}

// --- Since Topics ---

type SinceTopic struct {
	Topic         string `db:"topic"`
	SinceDatetime string `db:"since_datetime"`
	Count         int    `db:"count"`
}

func (d *DB) GetSinceTopic(topic string) (*SinceTopic, error) {
	var t SinceTopic
	err := d.db.Get(&t, `SELECT * FROM since_topics WHERE topic = ?`, topic)
	if err != nil {
		return nil, err
	}
	return &t, nil
}

func (d *DB) UpsertSinceTopic(topic string) error {
	now := time.Now().Format(time.RFC3339)
	_, err := d.db.Exec(
		`INSERT INTO since_topics (topic, since_datetime, count) VALUES (?, ?, 1) ON CONFLICT(topic) DO UPDATE SET count = count + 1, since_datetime = ?`,
		topic, now, now,
	)
	return err
}

func (d *DB) GetAllSinceTopics(limit int) ([]SinceTopic, error) {
	var topics []SinceTopic
	err := d.db.Select(&topics, `SELECT * FROM since_topics ORDER BY count DESC LIMIT ?`, limit)
	return topics, err
}

// --- Prism Words ---

func (d *DB) AddPrismWord(word string) error {
	now := time.Now().Format(time.RFC3339)
	_, err := d.db.Exec(
		`INSERT INTO prism_words (word, count, last_use) VALUES (?, 1, ?) ON CONFLICT(word) DO UPDATE SET count = count + 1, last_use = ?`,
		word, now, now,
	)
	return err
}

type PrismWord struct {
	Word    string `db:"word"`
	Count   int    `db:"count"`
	LastUse string `db:"last_use"`
}

func (d *DB) GetAllPrismWords() ([]PrismWord, error) {
	var words []PrismWord
	err := d.db.Select(&words, `SELECT * FROM prism_words ORDER BY count DESC`)
	return words, err
}

// --- Length Users ---

type LengthUser struct {
	UserID    int64  `db:"user_id"`
	Username  string `db:"username"`
	FirstName string `db:"first_name"`
	LastName  string `db:"last_name"`
	Length    int    `db:"length"`
}

func (u *LengthUser) DisplayName() string {
	if u.Username != "" {
		return "@" + u.Username
	}
	name := u.FirstName
	if u.LastName != "" {
		name += " " + u.LastName
	}
	if name == "" {
		return "unknown"
	}
	return name
}

func (d *DB) UpsertLengthUser(userID int64, username, firstName, lastName string, length int) error {
	_, err := d.db.Exec(
		`INSERT INTO length_users (user_id, username, first_name, last_name, length) VALUES (?, ?, ?, ?, ?)
		 ON CONFLICT(user_id) DO UPDATE SET username = ?, first_name = ?, last_name = ?, length = ?`,
		userID, username, firstName, lastName, length,
		username, firstName, lastName, length,
	)
	return err
}

func (d *DB) GetTopLengthUsers(limit int) ([]LengthUser, error) {
	var users []LengthUser
	err := d.db.Select(&users, `SELECT * FROM length_users ORDER BY length DESC LIMIT ?`, limit)
	return users, err
}

// --- Buktopuha Players ---

type BuktopuhaPlayer struct {
	UserID      int64  `db:"user_id"`
	Meta        string `db:"meta"`
	GameCounter int    `db:"game_counter"`
	WinCounter  int    `db:"win_counter"`
	TotalScore  int    `db:"total_score"`
	CreatedAt   string `db:"created_at"`
	UpdatedAt   string `db:"updated_at"`
}

func (p *BuktopuhaPlayer) Username() string {
	var meta map[string]any
	_ = json.Unmarshal([]byte(p.Meta), &meta)
	if username, ok := meta["username"].(string); ok && username != "" {
		return username
	}
	firstName, _ := meta["first_name"].(string)
	lastName, _ := meta["last_name"].(string)
	name := firstName
	if lastName != "" {
		if name != "" {
			name += " "
		}
		name += lastName
	}
	if name == "" {
		return "unknown"
	}
	return name
}

func (d *DB) FindBuktopuhaPlayer(userID int64) (*BuktopuhaPlayer, error) {
	var p BuktopuhaPlayer
	err := d.db.Get(&p, `SELECT * FROM buktopuha_players WHERE user_id = ?`, userID)
	if err != nil {
		return nil, err
	}
	return &p, nil
}

func (d *DB) AddBuktopuhaPlayer(userID int64, meta string, score int) error {
	now := time.Now().Format(time.RFC3339)
	winCounter := 0
	if score > 0 {
		winCounter = 1
	}
	_, err := d.db.Exec(
		`INSERT INTO buktopuha_players (user_id, meta, game_counter, win_counter, total_score, created_at, updated_at)
		 VALUES (?, ?, 1, ?, ?, ?, ?)`,
		userID, meta, winCounter, score, now, now,
	)
	return err
}

func (d *DB) IncBuktopuhaWin(userID int64, score int) error {
	now := time.Now().Format(time.RFC3339)
	_, err := d.db.Exec(
		`UPDATE buktopuha_players SET win_counter = win_counter + 1, total_score = total_score + ?, updated_at = ? WHERE user_id = ?`,
		score, now, userID,
	)
	return err
}

func (d *DB) IncBuktopuhaGameCounter(userID int64) error {
	now := time.Now().Format(time.RFC3339)
	_, err := d.db.Exec(
		`UPDATE buktopuha_players SET game_counter = game_counter + 1, updated_at = ? WHERE user_id = ?`,
		now, userID,
	)
	return err
}

func (d *DB) GetAllBuktopuhaPlayers() ([]BuktopuhaPlayer, error) {
	var players []BuktopuhaPlayer
	err := d.db.Select(&players, `SELECT * FROM buktopuha_players ORDER BY win_counter DESC`)
	return players, err
}

// --- AOC Data ---

func (d *DB) GetAOCData() (string, error) {
	var data string
	err := d.db.Get(&data, `SELECT data FROM aoc WHERE id = 0`)
	if err != nil {
		return "", err
	}
	return data, nil
}

func (d *DB) SetAOCData(data string) error {
	_, err := d.db.Exec(
		`INSERT INTO aoc (id, data) VALUES (0, ?) ON CONFLICT(id) DO UPDATE SET data = ?`,
		data, data,
	)
	return err
}
